### PR TITLE
Fix keras early stopping logging logic

### PR DESCRIPTION
**Pull request by harupy**
Signed-off-by: harupy <hkawamura0130@gmail.com>

## What changes are proposed in this pull request?

This PR fixes keras early stopping logging logic.

## How is this patch tested?

Unit tests

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

Tensorflow and keras autologging don't log `stopped_epoch` when training is not early stopped.

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [x] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/keras.py [hunk 1]
```diff
@@ -734,25 +734,31 @@ def _get_early_stop_callback_attrs(callback):
             return None
 
     def _log_early_stop_callback_metrics(callback, history, metrics_logger):
-        if callback:
-            callback_attrs = _get_early_stop_callback_attrs(callback)
-            if callback_attrs is None:
-                return
-            stopped_epoch, restore_best_weights, patience = callback_attrs
-            metrics_logger.record_metrics({"stopped_epoch": stopped_epoch})
```
Comment (harupy) [inline mlflow/keras.py:742]: The current implementation always logs `stopped_epoch` even when training didn't early stop, which might be confusing.

**Review by dbczumar — APPROVED** [review]
LGTM! Thanks @harupy !

File: mlflow/keras.py [hunk 2]
```diff
@@ -734,25 +734,31 @@ def _get_early_stop_callback_attrs(callback):
...
+            return
+
+        stopped_epoch, restore_best_weights, patience = callback_attrs
+        metrics_logger.record_metrics({"stopped_epoch": stopped_epoch})
+
+        restored_best_weights = restore_best_weights and callback.best_weights is not None
+        if not restored_best_weights:
+            return
+
+        restored_epoch = 0 if stopped_epoch == 0 else stopped_epoch - max(1, patience)
```
Comment (harupy) [inline mlflow/keras.py:751]:
```suggestion
        restored_epoch = history.history[callback.monitor].index(callback.best)
```

File: mlflow/tensorflow.py [hunk 3]
```diff
@@ -1050,39 +1050,50 @@ def _get_early_stop_callback_attrs(callback):
...
-                    if "loss" in restored_metrics:
-                        restored_metrics["epoch_loss"] = restored_metrics.pop("loss")
-                    if "acc" in restored_metrics:
-                        restored_metrics["epoch_acc"] = restored_metrics.pop("acc")
-                # Checking that a metric history exists
-                metric_key = next(iter(history.history), None)
-                if metric_key is not None:
-                    last_epoch = len(history.history[metric_key])
-                    metrics_logger.record_metrics(restored_metrics, last_epoch)
+    def _log_early_stop_callback_metrics(callback, history, metrics_logger):
```
Comment (harupy) [inline mlflow/tensorflow.py:1053]: `keras.py` and `tensorflow.py` contains several duplicate functions. We might want to move them in e.g. `utils.py`.

**Pull request merged**